### PR TITLE
pmemcheck: store trace support

### PR DESCRIPTION
**Pull request by wlemkows**
store trace/store trace depth

---
This change is [Reviewable](https://reviewable.io/reviews/pmem/valgrind/50)

### CODE REVIEW

File: coregrind/m_execontext.c [hunk 1]
```diff
@@ -7,7 +7,7 @@
    This file is part of Valgrind, a dynamic binary instrumentation
    framework.
 
-   Copyright (C) 2000-2015 Julian Seward 
+   Copyright (C) 2000-2015 Julian Seward
```
Comment (marcinslusarz) [inline coregrind/m_execontext.c]: Please don't touch unrelated lines when modifying things outside of pmemcheck directory. Such changes may be hard to rebase when updating pmemcheck to the next version.

File: coregrind/m_execontext.c [hunk 2]
```diff
@@ -521,6 +521,12 @@ ExeContext* VG_(null_ExeContext) (void)
    return null_ExeContext;
 }
 
+void VG_(apply_ExeContext)( void(*action)(UInt n, Addr ip, void *uu_opaque),
+                                    ExeContext* ec, UInt n_ips )
+{
+   VG_(apply_StackTrace)( action, NULL, ec->ips, n_ips );
+}
```
Comment (marcinslusarz) [inline coregrind/m_execontext.c]: Please add this in a separate commit (EDIT: if answer to the above question is "yes").

File: coregrind/m_execontext.c [hunk 3]
```diff
@@ -521,6 +521,12 @@ ExeContext* VG_(null_ExeContext) (void)
    return null_ExeContext;
 }
 
+void VG_(apply_ExeContext)( void(*action)(UInt n, Addr ip, void *uu_opaque),
```
Comment (marcinslusarz) [inline coregrind/m_execontext.c]: Please verify we really need this inside coregrind.
I think it can be implemented in pmemcheck using:
```
VG_(apply_StackTrace)( action, NULL, get_ExeContext_StackTrace(ec), n_ips );
```

File: pmemcheck/pmc_main.c [hunk 4]
```diff
@@ -1905,6 +1963,10 @@ pmc_print_usage(void)
             "                               line size default [no]\n"
             "    --tx-only=<yes|no>         turn on transaction only memory\n"
             "                               modifications default [no]\n"
+            "    --store-trace=<yes|no>     add store stacktraces in log-stores\n"
+            "                               mode default [no]\n"
+            "    --store-trace-depth=<uint> set store stacktrace depth\n"
+            "                               in log-stores mode default [1]\n"
```
Comment (marcinslusarz) [inline pmemcheck/pmc_main.c]: Maybe something like this?

```
--log-stores-stacktraces=<yes|no>  dump stracktrace with each logged store\n
                                   default [no]\n
--log-stores-stacktraces-depth=<uint> depth of logged stacktraces\n
                                   default [1]\n
```

Please move it near --log-stores option.

**Comment by wlemkows** [comment]
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions.

---

*[pmemcheck/pmc_main.c, line 1969 at r1](https://reviewable.io/reviews/pmem/valgrind/50#-LCTIgZKfoCRSWtYqjFR-r1-1969:-LD5_8IeAyOXAJv7bOvV:b-896fix) ([raw file](https://github.com/pmem/valgrind/blob/1ad04bb04005f90612547e0002db216bf0ee52e9/pmemcheck/pmc_main.c#L1969)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

Maybe something like this?

```
--log-stores-stacktraces=<yes|no>  dump stracktrace with each logged store\n
                                   default [no]\n
--log-stores-stacktraces-depth=<uint> depth of logged stacktraces\n
                                   default [1]\n
```

Please move it near --log-stores option.

</blockquote></details>

Done.

---

*[coregrind/m_execontext.c, line 10 at r1](https://reviewable.io/reviews/pmem/valgrind/50#-LCTIgZKfoCRSWtYqjFP-r1-10:-LD5VouhEpWwN8nb-S7r:b-896fix) ([raw file](https://github.com/pmem/valgrind/blob/1ad04bb04005f90612547e0002db216bf0ee52e9/coregrind/m_execontext.c#L10)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

Please don't touch unrelated lines when modifying things outside of pmemcheck directory. Such changes may be hard to rebase when updating pmemcheck to the next version.
</blockquote></details>

Done.

---

*[coregrind/m_execontext.c, line 524 at r1](https://reviewable.io/reviews/pmem/valgrind/50#-LCTIgZKfoCRSWtYqjFP-r1-524:-LD6GWfr55725k4ed7G8:b-896fix) ([raw file](https://github.com/pmem/valgrind/blob/1ad04bb04005f90612547e0002db216bf0ee52e9/coregrind/m_execontext.c#L524)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

Please verify we really need this inside coregrind.
I think it can be implemented in pmemcheck using:
```
VG_(apply_StackTrace)( action, NULL, get_ExeContext_StackTrace(ec), n_ips );
```
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/pmem/valgrind/50)*

**Comment by wlemkows** [comment]
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions.

---

*[coregrind/m_execontext.c, line 528 at r1](https://reviewable.io/reviews/pmem/valgrind/50#-LCTIgZKfoCRSWtYqjFP-r1-528:-LD6G_6e9664CHyOtqdn:beqo7v2) ([raw file](https://github.com/pmem/valgrind/blob/1ad04bb04005f90612547e0002db216bf0ee52e9/coregrind/m_execontext.c#L528)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

Please add this in a separate commit (EDIT: if answer to the above question is "yes").
</blockquote></details>

No changes needed in coregrind

---


*Comments from [Reviewable](https://reviewable.io/reviews/pmem/valgrind/50)*

**Comment by marcinslusarz** [comment]
Reviewed 1 of 3 files at r2.
Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[pmemcheck/pmc_main.c, line 1969 at r1](https://reviewable.io/reviews/pmem/valgrind/50#-LCTIgZKfoCRSWtYqjFR-r1-1969:-LD6j5rl2OeJ0ZzvDbA9:bog1rxn) ([raw file](https://github.com/pmem/valgrind/blob/1ad04bb04005f90612547e0002db216bf0ee52e9/pmemcheck/pmc_main.c#L1969)):*
<details><summary><i>Previously, wlemkows (Weronika Lewandowska) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Please also update pmc_process_cmd_line_option.

---

*[pmemcheck/pmc_main.c, line 370 at r2](https://reviewable.io/reviews/pmem/valgrind/50#-LD6k56dA-iHKOs4ixJA:-LD6k56dA-iHKOs4ixJB:b-dq349b) ([raw file](https://github.com/pmem/valgrind/blob/c9d09dca5d2e0a78275d733ec595bfa0a057ad28/pmemcheck/pmc_main.c#L370)):*
> ```C
>    do {
>       const HChar *buf = VG_(describe_IP)(ip, iipc);
>       if (VG_(clo_xml)) {
> ```

empty line between those 2, please

---

*[pmemcheck/pmc_main.c, line 374 at r2](https://reviewable.io/reviews/pmem/valgrind/50#-LD6jlLECC4cZ_NMSCDs:-LD6jlLECC4cZ_NMSCDt:b-kgx8n5) ([raw file](https://github.com/pmem/valgrind/blob/c9d09dca5d2e0a78275d733ec595bfa0a057ad28/pmemcheck/pmc_main.c#L374)):*
> ```C
>       } else {
>          VG_(emit)("%s", buf);
>       }
> ```

Please remove {} for single line statements. This comment also applies to pp_store_trace.

---

*[pmemcheck/pmc_main.c, line 378 at r2](https://reviewable.io/reviews/pmem/valgrind/50#-LD6kB0d-vwzI_axZblU:-LD6kB0d-vwzI_axZblV:b-dq349b) ([raw file](https://github.com/pmem/valgrind/blob/c9d09dca5d2e0a78275d733ec595bfa0a057ad28/pmemcheck/pmc_main.c#L378)):*
> ```C
>       // Increase n to show "at" for only one level.
>    } while (VG_(next_IIPC)(iipc));
>    VG_(delete_IIPC)(iipc);
> ```

empty line between those 2, please

---

*[pmemcheck/pmc_main.c, line 712 at r2](https://reviewable.io/reviews/pmem/valgrind/50#-LD6jPf6Ck4_hc9-k6JU:-LD6jPf6Ck4_hc9-k6JV:bqj1ojy) ([raw file](https://github.com/pmem/valgrind/blob/c9d09dca5d2e0a78275d733ec595bfa0a057ad28/pmemcheck/pmc_main.c#L712)):*
> ```C
>         if (pmem.store_traces)
>             pp_store_trace(store, pmem.store_traces_depth);
> 	}
> ```

tab->spaces

---


*Comments from [Reviewable](https://reviewable.io/reviews/pmem/valgrind/50)*

**Comment by wlemkows** [comment]
Review status: 1 of 2 files reviewed at latest revision, 5 unresolved discussions.

---

*[pmemcheck/pmc_main.c, line 1969 at r1](https://reviewable.io/reviews/pmem/valgrind/50#-LCTIgZKfoCRSWtYqjFR-r1-1969:-LD6o8PxB1WTmRC4r97b:b-896fix) ([raw file](https://github.com/pmem/valgrind/blob/1ad04bb04005f90612547e0002db216bf0ee52e9/pmemcheck/pmc_main.c#L1969)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

Please also update pmc_process_cmd_line_option.
</blockquote></details>

Done.

---

*[pmemcheck/pmc_main.c, line 370 at r2](https://reviewable.io/reviews/pmem/valgrind/50#-LD6k56dA-iHKOs4ixJA:-LD6llmNBdQPRVenVi81:b-896fix) ([raw file](https://github.com/pmem/valgrind/blob/c9d09dca5d2e0a78275d733ec595bfa0a057ad28/pmemcheck/pmc_main.c#L370)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

empty line between those 2, please
</blockquote></details>

Done.

---

*[pmemcheck/pmc_main.c, line 374 at r2](https://reviewable.io/reviews/pmem/valgrind/50#-LD6jlLECC4cZ_NMSCDs:-LD6lqUxEmVvkvOI-Ijz:b-896fix) ([raw file](https://github.com/pmem/valgrind/blob/c9d09dca5d2e0a78275d733ec595bfa0a057ad28/pmemcheck/pmc_main.c#L374)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

Please remove {} for single line statements. This comment also applies to pp_store_trace.
</blockquote></details>

Done.

---

*[pmemcheck/pmc_main.c, line 378 at r2](https://reviewable.io/reviews/pmem/valgrind/50#-LD6kB0d-vwzI_axZblU:-LD6mB5h7Maki1XhsTjp:b-896fix) ([raw file](https://github.com/pmem/valgrind/blob/c9d09dca5d2e0a78275d733ec595bfa0a057ad28/pmemcheck/pmc_main.c#L378)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

empty line between those 2, please
</blockquote></details>

Done.

---

*[pmemcheck/pmc_main.c, line 712 at r2](https://reviewable.io/reviews/pmem/valgrind/50#-LD6jPf6Ck4_hc9-k6JU:-LD6mOlwEL5IN3FfLLxR:b-896fix) ([raw file](https://github.com/pmem/valgrind/blob/c9d09dca5d2e0a78275d733ec595bfa0a057ad28/pmemcheck/pmc_main.c#L712)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

tab->spaces
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/pmem/valgrind/50)*

**Comment by marcinslusarz** [comment]
Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[pmemcheck/pmc_main.c, line 712 at r2](https://reviewable.io/reviews/pmem/valgrind/50#-LD6jPf6Ck4_hc9-k6JU:-LD6sVap5JoALJ3i_BlO:b-xaaest) ([raw file](https://github.com/pmem/valgrind/blob/c9d09dca5d2e0a78275d733ec595bfa0a057ad28/pmemcheck/pmc_main.c#L712)):*
<details><summary><i>Previously, wlemkows (Weronika Lewandowska) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Well, indentation is now a bit off.

---


*Comments from [Reviewable](https://reviewable.io/reviews/pmem/valgrind/50)*

**Comment by wlemkows** [comment]
Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[pmemcheck/pmc_main.c, line 712 at r2](https://reviewable.io/reviews/pmem/valgrind/50#-LD6jPf6Ck4_hc9-k6JU:-LD6twhU5eN6RHexIep4:bce861i) ([raw file](https://github.com/pmem/valgrind/blob/c9d09dca5d2e0a78275d733ec595bfa0a057ad28/pmemcheck/pmc_main.c#L712)):*
<details><summary><i>Previously, marcinslusarz (Marcin Ślusarz) wrote…</i></summary><blockquote>

Well, indentation is now a bit off.
</blockquote></details>

ups, done?

---


*Comments from [Reviewable](https://reviewable.io/reviews/pmem/valgrind/50)*

**Comment by marcinslusarz** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/pmem/valgrind/50#-:-LD7-N57BMO0JgdKpeV2:bnfp4nl)*
